### PR TITLE
WIP all: Add container types with disk overflow

### PR DESCRIPTION
**Pull request by imsodin**
### Purpose

Initial implementation of what I talked about in #4976. The idea is to restrict memory used in big slices/maps based on total available memory.

### Todo

 - Testing
  - Total memory on freebsd, openbsd and dragonfly is new, does it work?
 - Fallback if getting total memory fails.
 - Actually think about where to use these types instead of dropping it all over scanner and puller - which might be totally fine though :P

### Testing

Only existing ones so far, so only in-memory operations are tested, disk-spilling is untested.

### CODE REVIEW

**Comment by calmh** [comment]
Is this your code, or where does it come from?

**Comment by imsodin** [comment]
Very originally it was your index sorting code, but it became quite unrelated (I rewrote almost the entire thing a few times...). However https://github.com/syncthing/syncthing/pull/4980/files#diff-dbf2b1d1fa6344048d9bc7c2ce21be8aR167 is probably still the same (or I copied it from some lib/db file, don't remember), in which case I should keep 2016 as copyright date (?).

**Comment by calmh** [comment]
Oh alright I misunderstood your comment on the issue that was not original code, but then I get it.

File: lib/diskoverflow/sorted.go [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+		key:      lim.register(),
+		location: location,
+	}
+	s.commonSorted = &memorySorted{key: s.key}
+	return s
+}
+
+// SortValue must be implemented by every supported type for sorting. The sorting
+// will happen according to bytes.Compare on the key.
+type SortValue interface {
```
Comment (AudriusButkevicius) [inline lib/diskoverflow/sorted.go]: Why can't this just be
```
type SortValue interface {
  LessThan(other SortValue)
  Value
}
```

File: lib/diskoverflow/sorted.go [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+
+func (s *sorted) IterAndClose(fn func(v Value) bool, rev bool) {
+	s.iterImpl(fn, rev, true)
+	s.Close()
+}
+
+func (s *sorted) iterImpl(fn func(v Value) bool, rev, closing bool) {
+	if closing {
+		defer s.Close()
+	}
```
Comment (AudriusButkevicius) [inline lib/diskoverflow/sorted.go]: WHy is this needed if IterAndClose calls s.Close() anyway?

File: lib/diskoverflow/sorted.go [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+		close(iChan)
+	}()
+	av, aok := <-aChan
+	iv, iok := <-iChan
+	comp := -1
+	if rev {
+		comp = 1
+	}
+	for aok && iok {
+		if bytes.Compare(av.Key(), iv.Key()) == comp {
```
Comment (AudriusButkevicius) [inline lib/diskoverflow/sorted.go]: This could be insertion sort I guess. Also, this has the caveat that the key has to be unique, otherwise you lose values.

File: lib/diskoverflow/slice.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	orig := s.bytes()
+	removed := int64(0)
+	for i := 0; i < len(s.values); i++ {
+		if rev {
+			i = len(s.values) - 1 - i
+		}
+		if !fn(s.values[i]) {
+			return
+		}
+		if closing {
```
[the remark this answers is not in the document]
Comment (AudriusButkevicius) [reply lib/diskoverflow/slice.go]: I guess it would make sense if we were somehow appending as we iterate, but we never do that, so doing all of this stuff potentially slows things down, and doesn't alleviate memory pressure as I suspect it's unlikely for GC to be invoked while while we are dereferencing things.

File: lib/diskoverflow/sorted.go (hunk 1 above)
Comment (imsodin) [reply lib/diskoverflow/sorted.go]: I need a key to store it in the key-value-database.

File: lib/diskoverflow/slice.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+)
+
+type Slice interface {
+	Common
+	Append(v Value)
+	Bytes() int64 // Total estimated size of contents
+	Iter(fn func(v Value) bool, rev bool)
+	IterAndClose(fn func(v Value) bool, rev bool)
+}
+
```
[the remark this answers is not in the document]
Comment (imsodin) [reply lib/diskoverflow/slice.go]: I only implemented what's already needed, and indexing isn't. Can definitely be added.

File: lib/diskoverflow/map.go [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+package diskoverflow
+
+import (
+	"io/ioutil"
+	"os"
+
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+type Map interface {
```
[the remark this answers is not in the document]
Comment (imsodin) [reply lib/diskoverflow/map.go]: Probably you are right. It was just nice to have interfaces as a kind of planning board for what the types should be able to do. Also results in nice/early build errors.

File: lib/diskoverflow/slice.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (imsodin) [reply lib/diskoverflow/slice.go]: > [...] as I suspect it's unlikely for GC to be invoked while while we are dereferencing things.

E.g. scanning can take a long time, where I'd expect GC to kick in. Also huge sync operations (e.g. resolving conflicts on initial sync) do take a lot of time on a "normal" laptop. And all of this only happens when the container is bigger than 10MB.

File: lib/diskoverflow/sorted.go (hunk 3 above)
Comment (imsodin) [reply lib/diskoverflow/sorted.go]: Insertion sort is just a lot slower than the current insert sequentially and sort when a value is queried for the first time. And I indeed did overlook the uniqueness problem, but that's easily solvable by adding an insertion counter at the end of the key.

File: lib/diskoverflow/common.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+import (
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+)
+
+// Value must be implemented by every type that is to be stored in a disk spilling container.
+type Value interface {
+	Bytes() int64 // estimated size in memory in bytes
+	ToByte() []byte
+	FromByte([]byte)
```
[the remark this answers is not in the document]
Comment (imsodin) [reply lib/diskoverflow/common.go]: These encoding interfaces allow for failure, i.e. return an error. I really don't know how to handle an error gracefully and littering err!=nil->panic all over the place is ugly. Naming is definitely shit, Marshal, Unmarshal and Size is much better.

File: lib/diskoverflow/map.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+type commonMap interface {
+	common
+	add(k string, v Value)
+	get(k string) (Value, bool)
+	iter(fn func(k string, v Value) bool)
+	pop(k string) (Value, bool)
+}
+
+type intMap struct {
```
[the remark this answers is not in the document]
Comment (imsodin) [reply lib/diskoverflow/map.go]: I can't use `map` and my imagination didn't get any further than internal map, which became `intMap` due to lazyness. I am open to suggestions.

File: lib/diskoverflow/common.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+func (s ValueFileInfoSlice) ToByte() []byte {
+	data, err := s.Marshal()
+	if err != nil {
+		panic("bug: marshalling FileInfo should never fail: " + err.Error())
+	}
+	return data
+}
+
+func (s ValueFileInfoSlice) FromByte(v []byte) {
+	var f protocol.FileInfo
```
[the remark this answers is not in the document]
Comment (imsodin) [reply lib/diskoverflow/common.go]: Better in bep_extensions.go? It was a coin toss for me, because it's both diskoverflow specific (implemented interface) and protocol specific (underlying type).

File: lib/diskoverflow/map.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (AudriusButkevicius) [reply lib/diskoverflow/map.go]: Well, if there was less interfaces, you'd have Map available to you.

I don't see a purpose of these interfaces if one of them uses capitals the other one uses lowercase names...

**Comment by imsodin** [comment]
> -  I've grown less fond of the function based iterator and would prefer something like the iterators in leveldb - something you create, call Next() and Value() and so forth on in a loop, and then have a deferred Close() on. I think this makes for better code at the call site, maybe something to look into?

I implemented this "object style" iterators. Personally I don't even like the code on the call site. On the implementation side it's anyway pretty ugly (maybe I just don't know how to make it pretty). And I didn't implement the full functionality of e.g. the leveldb iterator (that would be a huge pain to do). What do you think? https://github.com/syncthing/syncthing/pull/4980/commits/986af92e986ddf7465f832200a88dffdb28f2af6

> - I don't think these types should concern themselves with memory sizing, instead just taking a maximum size at construction time. That size could potentially be something auto generated, but mostly I think it's good enough to have a configurable that defaults to something like 8 or 16 megs - just to avoid disk spilling for trivial operations.

I'd like to avoid disk spilling at all times. In my personal experience disk io is the weakest point of syncthing's performance - especially when huge sync operations happen. That's exactly when disk spilling of these types might also happen. So I really only want it to happen, if OOM is a danger. With a per object limit or with a limit that does not depend on system RAM, even on a system with plenty of RAM Syncthing will do more disk trashing.

**Comment by st-review** [comment]
:robot: *beep boop*

I'm going to close this pull request as it has been idle for more than 90 days.

This is not a rejection, merely a removal from the list of active pull requests that is periodically reviewed by humans. The pull request can be reopened when there is new activity, and merged once any remaining issues are resolved.
